### ISSUE TITLE
Blocking operation is not allowed in CRIU single thread mode

### ISSUE THREAD

**Issue by pshipton** [eclipse-openj9/openj9#18749]
[Internal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal/1)
cmdLineTester_criu_nonPortableRestore_4 `-Xjit:count=0 -XX:+CRIURestoreNonPortableMode`
ubu22-aarch64-6
```
Testing: Create and Restore Criu Checkpoint Image once - TestConcurrentModePostRestoreHookRunOnce
Test start time: 2024/01/11 16:01:18 Eastern Standard Time
Running command: bash /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/jdkbinary/j2sdk-image/bin/java " -Xjit:count=0 -XX:+CRIURestoreNonPortableMode " org.openj9.criu.TestConcurrentMode TestConcurrentModePostRestoreHookRunOnce 1 false false
Time spent starting: 3 milliseconds
Time spent executing: 4921 milliseconds
Test result: FAILED
Output from test:
 [OUT] start running script
 [OUT] export GLIBC_TUNABLES=glibc.cpu.hwcaps=-XSAVEC,-XSAVE,-AVX2,-ERMS,-AVX,-AVX_Fast_Unaligned_Load
 [OUT] export LD_BIND_NOT=on
 [OUT] /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/jdkbinary/j2sdk-image/bin/java -XX:+EnableCRIUSupport  -Xjit:count=0 -XX:+CRIURestoreNonPortableMode  -cp /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criu.jar org.openj9.criu.TestConcurrentMode TestConcurrentModePostRestoreHookRunOnce 1
 [OUT] TestConcurrentModePostRestoreHookRunOnce() starts .., current thread name: main, Thu Jan 11 16:01:19 EST 2024, System.currentTimeMillis(): 1705006879469, System.nanoTime(): 218285649887882
 [OUT] TestConcurrentModePostRestoreHookRunOnce() Pre-checkpoint, current thread name: main, Thu Jan 11 16:01:21 EST 2024, System.currentTimeMillis(): 1705006881157, System.nanoTime(): 218287338258382
 [OUT] Performing CRIUSupport.checkpointJVM(), current thread name: main, Thu Jan 11 16:01:21 EST 2024, System.currentTimeMillis(): 1705006881158, System.nanoTime(): 218287339817115
 [OUT] JVMJITM044W Some or all compiled code in the code cache invalidated post restore.
 [OUT] JVMDUMP034I User requested Java dump using '/home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/aqa-tests/TKG/output_17050062176517/cmdLineTester_criu_nonPortableRestore_4/javacore.20240111.160123.25386.0001.txt' through CRIUSingleThreadModeJVMCRIUException
 [OUT] JVMDUMP010I Java dump written to /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/aqa-tests/TKG/output_17050062176517/cmdLineTester_criu_nonPortableRestore_4/javacore.20240111.160123.25386.0001.txt
 [OUT] Exception in thread "main" java.lang.BootstrapMethodError: openj9.internal.criu.JVMRestoreException: Blocking operation is not allowed in CRIU single thread mode.
 [OUT] 	at java.base/openj9.internal.criu.InternalCRIUSupport.clearInetAddressCache(InternalCRIUSupport.java:825)
 [OUT] 	at java.base/openj9.internal.criu.J9InternalCheckpointHookAPI$J9InternalCheckpointHook.runHook(J9InternalCheckpointHookAPI.java:143)
 [OUT] 	at java.base/openj9.internal.criu.J9InternalCheckpointHookAPI.runHooks(J9InternalCheckpointHookAPI.java:98)
 [OUT] 	at java.base/openj9.internal.criu.J9InternalCheckpointHookAPI.runPostRestoreHooksSingleThread(J9InternalCheckpointHookAPI.java:115)
 [OUT] 	at java.base/openj9.internal.criu.InternalCRIUSupport.checkpointJVMImpl(Native Method)
 [OUT] 	at java.base/openj9.internal.criu.InternalCRIUSupport.checkpointJVM(InternalCRIUSupport.java:880)
 [OUT] 	at openj9.criu/org.eclipse.openj9.criu.CRIUSupport.checkpointJVM(CRIUSupport.java:511)
 [OUT] 	at org.openj9.criu.CRIUTestUtils.checkPointJVMNoSetup(CRIUTestUtils.java:103)
 [OUT] 	at org.openj9.criu.TestConcurrentMode.TestConcurrentModePostRestoreHookRunOnce(TestConcurrentMode.java:208)
 [OUT] 	at org.openj9.criu.TestConcurrentMode.main(TestConcurrentMode.java:63)
 [OUT] Caused by: openj9.internal.criu.JVMRestoreException: Blocking operation is not allowed in CRIU single thread mode.
 [OUT] 	at java.base/jdk.internal.ref.PhantomCleanable.insert(PhantomCleanable.java:87)
 [OUT] 	at java.base/jdk.internal.ref.PhantomCleanable.<init>(PhantomCleanable.java:68)
 [OUT] 	at java.base/jdk.internal.ref.CleanerImpl$PhantomCleanableRef.<init>(CleanerImpl.java:172)
 [OUT] 	at java.base/java.lang.ref.Cleaner.register(Cleaner.java:220)
 [OUT] 	at java.base/java.lang.invoke.MethodHandleNatives$CallSiteContext.make(MethodHandleNatives.java:86)
 [OUT] 	at java.base/java.lang.invoke.CallSite.<init>(CallSite.java:143)
 [OUT] 	at java.base/java.lang.invoke.ConstantCallSite.<init>(ConstantCallSite.java:44)
 [OUT] 	at java.base/java.lang.invoke.InnerClassLambdaMetafactory.buildCallSite(InnerClassLambdaMetafactory.java:219)
 [OUT] 	at java.base/java.lang.invoke.LambdaMetafactory.metafactory(LambdaMetafactory.java:329)
 [OUT] 	at java.base/java.lang.invoke.BootstrapMethodInvoker.invoke(BootstrapMethodInvoker.java:127)
 [OUT] 	at java.base/java.lang.invoke.CallSite.makeSite(CallSite.java:307)
 [OUT] 	at java.base/java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:258)
 [OUT] 	at java.base/java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:248)
 [OUT] 	at java.base/java.lang.invoke.MethodHandleResolver.resolveInvokeDynamic(MethodHandleResolver.java:194)
 [OUT] 	... 10 more
 [OUT] Removed test output files
 [OUT] finished script
 [ERR] /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_aarch64_linux_ojdk292_Personal_testList_0/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh: line 41: 25386 Killed                  $2 -XX:+EnableCRIUSupport $3 -cp "$1/criu.jar" $4 $5 $6 > testOutput 2>&1
>> Success condition was not found: [Output match: TestConcurrentModePostRestoreHookRunOnce() within postRestoreHook]
>> Required condition was found: [Output match: Pre-checkpoint]
>> Required condition was found: [Output match: Killed]
>> Failure condition was not found: [Output match: CRIU is not enabled]
>> Failure condition was not found: [Output match: Operation not permitted]
>> Success condition was not found: [Output match: Thread pid mismatch]
>> Success condition was not found: [Output match: do not match expected]
>> Success condition was not found: [Output match: Unable to create a thread:]
>> Failure condition was not found: [Output match: Could not dump the JVM processes, err=-70]
>> Failure condition was found: [Output match: User requested Java dump using]
```

**Comment by pshipton** [comment]
@TobiAjila @singh264 @babsingh fyi

**Comment by babsingh** [comment]
This might be a duplicate of #18670. It is run with `-Xjit:count=0`. If we look at the native stack, we might be indefinitely waiting on the `JIT-QueueSlotMonitor`. @singh264 Can you please confirm?

**Comment by singh264** [comment]
> This might be a duplicate of https://github.com/eclipse-openj9/openj9/issues/18670. It is run with -Xjit:count=0. In the native stack, we might be indefinitely waiting on the JIT-QueueSlotMonitor. @singh264 Can you please confirm?

This seems to be a duplicate of https://github.com/eclipse-openj9/openj9/issues/18399 where a thread is halted while it holds a shared resource in [PhantomCleanable](https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/d89d744ca980f422c89e680b45eb662a3c9023f4/src/java.base/share/classes/jdk/internal/ref/PhantomCleanable.java#L87), and the checkpoint thread is unable to acquire the same shared resource during restore in single thread mode. A solution was added to jdk17, jdk21, jdk-next, but not to jdk11 as no failures occurred when running  the tests in `cmdLineTester_criu_nonPortableRestore` locally on a `x86-64 Linux` 500 times.

**Comment by pshipton** [comment]
The failure is with OpenJDK method handles enabled, which we plan to make the default in the future. The feature has changed what code is run and therefore the timing of things as well. We should backport the fix to jdk11.

**Comment by singh264** [comment]
A [grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/37365/) running `TestConcurrentModePostRestoreHookRunOnce` 1000 times on 2 `aarch64 linux` machines passed with a openjdk method handles enabled build that included [NotCheckpointSafe annotations in PhantomCleanable](https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/750). 

Re-running the [grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/37363/) without NotCheckpointSafe annotations in PhantomCleanable reproduced the test failure where the checkpoint thread was blocked at [`PhantomCleanable.insert`](https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/d89d744ca980f422c89e680b45eb662a3c9023f4/src/java.base/share/classes/jdk/internal/ref/PhantomCleanable.java#L87) during restore in single thread mode, and the `Common-Cleaner` thread was holding the shared resource at [`PhantomCleanable.isListEmpty`](https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/d89d744ca980f422c89e680b45eb662a3c9023f4/src/java.base/share/classes/jdk/internal/ref/PhantomCleanable.java#L121).

```
[2024-01-13T04:55:48.273Z] TESTING:
[2024-01-13T04:55:49.204Z] *** Starting test suite: J9 Criu Command-Line Option Tests ***
[2024-01-13T04:55:49.204Z] Testing: Create and Restore Criu Checkpoint Image once - TestConcurrentModePostRestoreHookRunOnce
[2024-01-13T04:55:49.634Z] Test start time: 2024/01/12 23:55:49 Eastern Standard Time
[2024-01-13T04:55:50.063Z] Running command: bash /home/jenkins/workspace/Grinder_iteration_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh /home/jenkins/workspace/Grinder_iteration_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu /home/jenkins/workspace/Grinder_iteration_1/jdkbinary/j2sdk-image/bin/java " -Xjit:count=0 -XX:+CRIURestoreNonPortableMode -Xdump:java+system+jit:events=systhrow,filter=openj9/internal/criu/JVMRestoreException" org.openj9.criu.TestConcurrentMode TestConcurrentModePostRestoreHookRunOnce 1 false false
[2024-01-13T04:55:50.063Z] Time spent starting: 182 milliseconds
[2024-01-13T04:56:50.186Z] Time spent executing: 57035 milliseconds
[2024-01-13T04:56:50.186Z] Test result: FAILED
[2024-01-13T04:56:50.186Z] Output from test:
[2024-01-13T04:56:50.186Z]  [OUT] start running script
[2024-01-13T04:56:50.186Z]  [OUT] export GLIBC_TUNABLES=glibc.cpu.hwcaps=-XSAVEC,-XSAVE,-AVX2,-ERMS,-AVX,-AVX_Fast_Unaligned_Load
[2024-01-13T04:56:50.186Z]  [OUT] export LD_BIND_NOT=on
[2024-01-13T04:56:50.186Z]  [OUT] /home/jenkins/workspace/Grinder_iteration_1/jdkbinary/j2sdk-image/bin/java -XX:+EnableCRIUSupport  -Xjit:count=0 -XX:+CRIURestoreNonPortableMode -Xdump:java+system+jit:events=systhrow,filter=openj9/internal/criu/JVMRestoreException -cp /home/jenkins/workspace/Grinder_iteration_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criu.jar org.openj9.criu.TestConcurrentMode TestConcurrentModePostRestoreHookRunOnce 1
[2024-01-13T04:56:50.186Z]  [OUT] TestConcurrentModePostRestoreHookRunOnce() starts .., current thread name: main, Fri Jan 12 23:56:14 EST 2024, System.currentTimeMillis(): 1705121774289, System.nanoTime(): 333184742495781
[2024-01-13T04:56:50.186Z]  [OUT] TestConcurrentModePostRestoreHookRunOnce() Pre-checkpoint, current thread name: main, Fri Jan 12 23:56:32 EST 2024, System.currentTimeMillis(): 1705121792226, System.nanoTime(): 333202679674899
[2024-01-13T04:56:50.186Z]  [OUT] Performing CRIUSupport.checkpointJVM(), current thread name: main, Fri Jan 12 23:56:32 EST 2024, System.currentTimeMillis(): 1705121792233, System.nanoTime(): 333202686257283
...
[2024-01-13T04:56:50.187Z]  [OUT] Exception in thread "main" java.lang.BootstrapMethodError: openj9.internal.criu.JVMRestoreException: Blocking operation is not allowed in CRIU single thread mode.
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/openj9.internal.criu.InternalCRIUSupport.clearInetAddressCache(InternalCRIUSupport.java:825)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/openj9.internal.criu.J9InternalCheckpointHookAPI$J9InternalCheckpointHook.runHook(J9InternalCheckpointHookAPI.java:143)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/openj9.internal.criu.J9InternalCheckpointHookAPI.runHooks(J9InternalCheckpointHookAPI.java:98)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/openj9.internal.criu.J9InternalCheckpointHookAPI.runPostRestoreHooksSingleThread(J9InternalCheckpointHookAPI.java:115)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/openj9.internal.criu.InternalCRIUSupport.checkpointJVMImpl(Native Method)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/openj9.internal.criu.InternalCRIUSupport.checkpointJVM(InternalCRIUSupport.java:880)
[2024-01-13T04:56:50.187Z]  [OUT] 	at openj9.criu/org.eclipse.openj9.criu.CRIUSupport.checkpointJVM(CRIUSupport.java:511)
[2024-01-13T04:56:50.187Z]  [OUT] 	at org.openj9.criu.CRIUTestUtils.checkPointJVMNoSetup(CRIUTestUtils.java:103)
[2024-01-13T04:56:50.187Z]  [OUT] 	at org.openj9.criu.TestConcurrentMode.TestConcurrentModePostRestoreHookRunOnce(TestConcurrentMode.java:208)
[2024-01-13T04:56:50.187Z]  [OUT] 	at org.openj9.criu.TestConcurrentMode.main(TestConcurrentMode.java:63)
[2024-01-13T04:56:50.187Z]  [OUT] Caused by: openj9.internal.criu.JVMRestoreException: Blocking operation is not allowed in CRIU single thread mode.
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/jdk.internal.ref.PhantomCleanable.insert(PhantomCleanable.java:87)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/jdk.internal.ref.PhantomCleanable.<init>(PhantomCleanable.java:68)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/jdk.internal.ref.CleanerImpl$PhantomCleanableRef.<init>(CleanerImpl.java:172)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.ref.Cleaner.register(Cleaner.java:220)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.MethodHandleNatives$CallSiteContext.make(MethodHandleNatives.java:86)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.CallSite.<init>(CallSite.java:143)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.ConstantCallSite.<init>(ConstantCallSite.java:44)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.InnerClassLambdaMetafactory.buildCallSite(InnerClassLambdaMetafactory.java:219)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.LambdaMetafactory.metafactory(LambdaMetafactory.java:329)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.BootstrapMethodInvoker.invoke(BootstrapMethodInvoker.java:127)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.CallSite.makeSite(CallSite.java:307)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:258)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:248)
[2024-01-13T04:56:50.187Z]  [OUT] 	at java.base/java.lang.invoke.MethodHandleResolver.resolveInvokeDynamic(MethodHandleResolver.java:194)
[2024-01-13T04:56:50.187Z]  [OUT] 	... 10 more
[2024-01-13T04:56:50.187Z]  [OUT] Removed test output files
[2024-01-13T04:56:50.187Z]  [OUT] finished script
[2024-01-13T04:56:50.187Z]  [ERR] /home/jenkins/workspace/Grinder_iteration_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh: line 41: 3738201 Killed                  $2 -XX:+EnableCRIUSupport $3 -cp "$1/criu.jar" $4 $5 $6 > testOutput 2>&1
[2024-01-13T04:56:50.187Z] >> Success condition was not found: [Output match: TestConcurrentModePostRestoreHookRunOnce() within postRestoreHook]
[2024-01-13T04:56:50.187Z] >> Required condition was found: [Output match: Pre-checkpoint]
[2024-01-13T04:56:50.187Z] >> Required condition was found: [Output match: Killed]
[2024-01-13T04:56:50.187Z] >> Failure condition was not found: [Output match: CRIU is not enabled]
[2024-01-13T04:56:50.187Z] >> Failure condition was not found: [Output match: Operation not permitted]
[2024-01-13T04:56:50.187Z] >> Success condition was not found: [Output match: Thread pid mismatch]
[2024-01-13T04:56:50.187Z] >> Success condition was not found: [Output match: do not match expected]
[2024-01-13T04:56:50.187Z] >> Success condition was not found: [Output match: Unable to create a thread:]
[2024-01-13T04:56:50.187Z] >> Failure condition was not found: [Output match: Could not dump the JVM processes, err=-70]
[2024-01-13T04:56:50.187Z] >> Failure condition was found: [Output match: User requested Java dump using]

> !threads
	!stack 0x00019200	!j9vmthread 0x00019200	!j9thread 0xffffb801c3c0	tid 0x390a5a (3738202) // (main) <-----
	!stack 0x0001b700	!j9vmthread 0x0001b700	!j9thread 0xffffb801cdc0	tid 0x390a5d (3738205) // (JIT Compilation Thread-000 Suspended)
	!stack 0x0001dc00	!j9vmthread 0x0001dc00	!j9thread 0xffffb81018b0	tid 0x390a5e (3738206) // (JIT Compilation Thread-001 Suspended)
	!stack 0x00020100	!j9vmthread 0x00020100	!j9thread 0xffffb8101db0	tid 0x390a5f (3738207) // (JIT Compilation Thread-002 Suspended)
	!stack 0x00022600	!j9vmthread 0x00022600	!j9thread 0xffffb81022b0	tid 0x390a60 (3738208) // (JIT Compilation Thread-003 Suspended)
	!stack 0x00024b00	!j9vmthread 0x00024b00	!j9thread 0xffffb8105e90	tid 0x390a61 (3738209) // (JIT Compilation Thread-004 Suspended)
	!stack 0x00027000	!j9vmthread 0x00027000	!j9thread 0xffffb8106390	tid 0x390a62 (3738210) // (JIT Compilation Thread-005)
	!stack 0x00029500	!j9vmthread 0x00029500	!j9thread 0xffffb8106890	tid 0x390a63 (3738211) // (JIT Compilation Thread-006 Suspended)
	!stack 0x0002ba00	!j9vmthread 0x0002ba00	!j9thread 0xffffb8107650	tid 0x390a64 (3738212) // (JIT Compilation Thread-007 Suspended)
	!stack 0x0002de00	!j9vmthread 0x0002de00	!j9thread 0xffffb8107b50	tid 0x390a65 (3738213) // (JIT Compilation Thread-008 Suspended)
	!stack 0x00030300	!j9vmthread 0x00030300	!j9thread 0xffffb8108050	tid 0x390a66 (3738214) // (JIT Compilation Thread-009 Suspended)
	!stack 0x00032800	!j9vmthread 0x00032800	!j9thread 0xffffb8108e10	tid 0x390a67 (3738215) // (JIT Compilation Thread-010 Suspended)
	!stack 0x00034d00	!j9vmthread 0x00034d00	!j9thread 0xffffb8109310	tid 0x390a68 (3738216) // (JIT Compilation Thread-011 Suspended)
	!stack 0x00037200	!j9vmthread 0x00037200	!j9thread 0xffffb8109810	tid 0x390a69 (3738217) // (JIT Compilation Thread-012 Suspended)
	!stack 0x00039700	!j9vmthread 0x00039700	!j9thread 0xffffb810a5d0	tid 0x390a6a (3738218) // (JIT Compilation Thread-013 Suspended)
	!stack 0x0003bc00	!j9vmthread 0x0003bc00	!j9thread 0xffffb810aad0	tid 0x390a6b (3738219) // (JIT Compilation Thread-014 Suspended)
	!stack 0x0003e100	!j9vmthread 0x0003e100	!j9thread 0xffffb810afd0	tid 0x390a6c (3738220) // (JIT Diagnostic Compilation Thread-015 Suspended)
	!stack 0x00040600	!j9vmthread 0x00040600	!j9thread 0xffffb8132970	tid 0x390a6d (3738221) // (JIT-SamplerThread)
	!stack 0x00129400	!j9vmthread 0x00129400	!j9thread 0xffffb8132e70	tid 0x390a77 (3738231) // (Common-Cleaner) <-----
	!stack 0x0013b900	!j9vmthread 0x0013b900	!j9thread 0xffffb824c7a0	tid 0x390a7d (3738237) // (Concurrent Mark Helper)
	!stack 0x0013de00	!j9vmthread 0x0013de00	!j9thread 0xffffb824cca0	tid 0x390a7e (3738238) // (GC Worker)
	!stack 0x00140300	!j9vmthread 0x00140300	!j9thread 0xffffb824d1a0	tid 0x390a7f (3738239) // (GC Worker)
	!stack 0x00142800	!j9vmthread 0x00142800	!j9thread 0xffffb824dba0	tid 0x390a80 (3738240) // (GC Worker)
	!stack 0x00187000	!j9vmthread 0x00187000	!j9thread 0xffff1000f980	tid 0x390a88 (3738248) // (Attach API wait loop)
	!stack 0x0015e100	!j9vmthread 0x0015e100	!j9thread 0xffffb824f9a0	tid 0x390a91 (3738257) // (Finalizer thread)
	!stack 0x00144d00	!j9vmthread 0x00144d00	!j9thread 0xffffb824efa0	tid 0x390a92 (3738258) // (GC Worker)
	!stack 0x0014bc00	!j9vmthread 0x0014bc00	!j9thread 0xffffb824f4a0	tid 0x390a93 (3738259) // (GC Worker)
	!stack 0x00149700	!j9vmthread 0x00149700	!j9thread 0xffffb824e5a0	tid 0x390a94 (3738260) // (GC Worker)
	!stack 0x00147200	!j9vmthread 0x00147200	!j9thread 0xffffb824e0a0	tid 0x390a95 (3738261) // (GC Worker)

> !stack 0x00019200 // (main)
<19200> 	                        Generic special frame
<19200> 	!j9method 0x000000000011C4F0   jdk/internal/ref/PhantomCleanable.insert()V <-----
<19200> 	!j9method 0x000000000011C4B0   jdk/internal/ref/PhantomCleanable.<init>(Ljava/lang/Object;Ljava/lang/ref/Cleaner;)V
<19200> 	!j9method 0x000000000011C9D8   jdk/internal/ref/CleanerImpl$PhantomCleanableRef.<init>(Ljava/lang/Object;Ljava/lang/ref/Cleaner;Ljava/lang/Runnable;)V
<19200> 	!j9method 0x000000000011AC70   java/lang/ref/Cleaner.register(Ljava/lang/Object;Ljava/lang/Runnable;)Ljava/lang/ref/Cleaner$Cleanable;
<19200> 	!j9method 0x0000000000207FB8   java/lang/invoke/MethodHandleNatives$CallSiteContext.make(Ljava/lang/invoke/CallSite;)Ljava/lang/invoke/MethodHandleNatives$CallSiteContext;
<19200> 	!j9method 0x00000000001B8328   java/lang/invoke/CallSite.<init>(Ljava/lang/invoke/MethodHandle;)V
<19200> 	!j9method 0x0000000000207218   java/lang/invoke/ConstantCallSite.<init>(Ljava/lang/invoke/MethodHandle;)V
<19200> 	!j9method 0x000000000022CB00   java/lang/invoke/InnerClassLambdaMetafactory.buildCallSite()Ljava/lang/invoke/CallSite;
<19200> 	!j9method 0x000000000022A300   java/lang/invoke/LambdaMetafactory.metafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite;
<19200> 	!j9method 0x00000000001F8788   java/lang/invoke/LambdaForm$DMH/0x0000000000000000.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
<19200> 	!j9method 0x0000000000232C98   java/lang/invoke/LambdaForm$MH/0x0000000000000000.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
<19200> 	!j9method 0x00000000001D71E8   java/lang/invoke/BootstrapMethodInvoker.invoke(Ljava/lang/Class;Ljava/lang/invoke/MethodHandle;Ljava/lang/String;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Class;)Ljava/lang/Object;
<19200> 	!j9method 0x00000000001B8548   java/lang/invoke/CallSite.makeSite(Ljava/lang/invoke/MethodHandle;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/Object;Ljava/lang/Class;)Ljava/lang/invoke/CallSite;
<19200> 	!j9method 0x0000000000184AA8   java/lang/invoke/MethodHandleNatives.linkCallSiteImpl(Ljava/lang/Class;Ljava/lang/invoke/MethodHandle;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/invoke/MemberName;
<19200> 	!j9method 0x0000000000184A88   java/lang/invoke/MethodHandleNatives.linkCallSite(Ljava/lang/Object;ILjava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/invoke/MemberName;
<19200> 	!j9method 0x00000000000936F8   java/lang/invoke/MethodHandleResolver.resolveInvokeDynamic(JLjava/lang/String;Ljava/lang/String;J)Ljava/lang/Object;
<19200> 	                        JNI call-in frame
<19200> 	                        known but unhandled frame type com.ibm.j9ddr.vm29.pointer.U8Pointer @ 0x00000005 
 FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT  FAULT 
<19200> 	!j9method 0x000000000024DAB8   openj9/internal/criu/InternalCRIUSupport.clearInetAddressCache()V
<19200> 	!j9method 0x0000000000261A48   openj9/internal/criu/InternalCRIUSupport$$Lambda$9/0x0000000000000000.run()V
<19200> 	!j9method 0x0000000000259E18   openj9/internal/criu/J9InternalCheckpointHookAPI$J9InternalCheckpointHook.runHook()V
<19200> 	!j9method 0x0000000000256140   openj9/internal/criu/J9InternalCheckpointHookAPI.runHooks(Ljava/util/List;Z)V
<19200> 	!j9method 0x0000000000256180   openj9/internal/criu/J9InternalCheckpointHookAPI.runPostRestoreHooksSingleThread()V
<19200> 	                        JNI call-in frame
<19200> 	!j9method 0x000000000024D6F8   openj9/internal/criu/InternalCRIUSupport.checkpointJVMImpl(Ljava/lang/String;ZZZILjava/lang/String;ZLjava/lang/String;ZZZZLjava/lang/String;Ljava/lang/String;)V
<19200> 	!j9method 0x000000000024DAD8   openj9/internal/criu/InternalCRIUSupport.checkpointJVM()V
<19200> 	!j9method 0x000000000024D248   org/eclipse/openj9/criu/CRIUSupport.checkpointJVM()V
<19200> 	!j9method 0x00000000001B13A8   org/openj9/criu/CRIUTestUtils.checkPointJVMNoSetup(Lorg/eclipse/openj9/criu/CRIUSupport;Ljava/nio/file/Path;Z)V
<19200> 	!j9method 0x00000000001B06A8   org/openj9/criu/TestConcurrentMode.TestConcurrentModePostRestoreHookRunOnce()V
<19200> 	!j9method 0x00000000001B05C8   org/openj9/criu/TestConcurrentMode.main([Ljava/lang/String;)V
<19200> 	                        JNI call-in frame
<19200> 	                        Native method frame
 
> !stackslots 0x00019200 // (main)
<19200> *** BEGIN STACK WALK, flags = 00400001 walkThread = 102,912 ***
<19200> 	ITERATE_O_SLOTS
<19200> 	RECORD_BYTECODE_PC_OFFSET
<19200> Initial values: walkSP = 0x00000000000F6B08, PC = 0x0000000000000001, literals = 0x0000000000000000, A0 = 0x00000000000F6B20, j2iFrame = 0x00000000000F6C48, decomp = 0x0000000000000000
<19200> Generic special frame: bp = 0x00000000000F6B20, sp = 0x00000000000F6B08, pc = 0x0000000000000001, cp = 0x0000000000000000, arg0EA = 0x00000000000F6B20, flags = 0x0000000000000000
<19200> Bytecode frame: bp = 0x00000000000F6B38, sp = 0x00000000000F6B28, pc = 0x0000FFFF395B5F72, cp = 0x000000000011C310, arg0EA = 0x00000000000F6B50, flags = 0x0000000000000000
<19200> 	Method: jdk/internal/ref/PhantomCleanable.insert()V !j9method 0x000000000011C4F0
<19200> 	Bytecode index = 6 <-----
<19200> 	Using local mapper
<19200> 	Locals starting at 0x00000000000F6B50 for 0x0000000000000003 slots
<19200> 		O-Slot: a0[0x00000000000F6B50] = 0x00000000FFE5D030
<19200> 		O-Slot: t1[0x00000000000F6B48] = 0x00000000899EA638
<19200> 		I-Slot: t2[0x00000000000F6B40] = 0x00000000000F6B80
...

> !bytecodes 0x000000000011C4F0
...
    0 aload0getfield 
    1 getfield 10 jdk/internal/ref/PhantomCleanable.list Ljdk/internal/ref/PhantomCleanable;
    4 dup 
    5 astore1 
    6 monitorenter <-----
    7 aload0 
    8 aload0getfield 
    9 getfield 10 jdk/internal/ref/PhantomCleanable.list Ljdk/internal/ref/PhantomCleanable;
   12 putfield 7 jdk/internal/ref/PhantomCleanable.prev Ljdk/internal/ref/PhantomCleanable;
   15 aload0 
   16 aload0getfield 
   17 getfield 10 jdk/internal/ref/PhantomCleanable.list Ljdk/internal/ref/PhantomCleanable;
   20 getfield 8 jdk/internal/ref/PhantomCleanable.next Ljdk/internal/ref/PhantomCleanable;
   23 putfield 8 jdk/internal/ref/PhantomCleanable.next Ljdk/internal/ref/PhantomCleanable;
   26 aload0getfield 
   27 getfield 8 jdk/internal/ref/PhantomCleanable.next Ljdk/internal/ref/PhantomCleanable;
   30 aload0 
   31 putfield 7 jdk/internal/ref/PhantomCleanable.prev Ljdk/internal/ref/PhantomCleanable;
   34 aload0getfield 
   35 getfield 10 jdk/internal/ref/PhantomCleanable.list Ljdk/internal/ref/PhantomCleanable;
   38 aload0 
   39 putfield 8 jdk/internal/ref/PhantomCleanable.next Ljdk/internal/ref/PhantomCleanable;
   42 aload1 
   43 monitorexit 
   44 goto 52
   47 astore2 
   48 aload1 
   49 monitorexit 
   50 aload2 
   51 athrow 
   52 return0 
...

> !stack 0x00129400 // (Common-Cleaner)
<129400> 	                        Generic special frame
<129400> 	!j9method 0x000000000011C530   jdk/internal/ref/PhantomCleanable.isListEmpty()Z <-----
<129400> 	!j9method 0x000000000011BD50   jdk/internal/ref/CleanerImpl.run()V
<129400> 	!j9method 0x0000000000061888   java/lang/Thread.run()V
<129400> 	!j9method 0x000000000011ED40   jdk/internal/misc/InnocuousThread.run()V
<129400> 	                        JNI call-in frame
<129400> 	                        Native method frame
 
> !stackslots 0x00129400 // (Common-Cleaner)
<129400> *** BEGIN STACK WALK, flags = 00400001 walkThread = 1,217,536 ***
<129400> 	ITERATE_O_SLOTS
<129400> 	RECORD_BYTECODE_PC_OFFSET
<129400> Initial values: walkSP = 0x00000000001291C8, PC = 0x0000000000000001, literals = 0x0000000000000000, A0 = 0x00000000001291E0, j2iFrame = 0x0000000000000000, decomp = 0x0000000000000000
<129400> Generic special frame: bp = 0x00000000001291E0, sp = 0x00000000001291C8, pc = 0x0000000000000001, cp = 0x0000000000000000, arg0EA = 0x00000000001291E0, flags = 0x0000000000000000
<129400> Bytecode frame: bp = 0x00000000001291F8, sp = 0x00000000001291E8, pc = 0x0000FFFF395B60AD, cp = 0x000000000011C310, arg0EA = 0x0000000000129210, flags = 0x0000000000000000
<129400> 	Method: jdk/internal/ref/PhantomCleanable.isListEmpty()Z !j9method 0x000000000011C530
<129400> 	Bytecode index = 25 <-----
<129400> 	Using local mapper
<129400> 	Locals starting at 0x0000000000129210 for 0x0000000000000003 slots
<129400> 		I-Slot: a0[0x0000000000129210] = 0x00000000899EA638
<129400> 		O-Slot: t1[0x0000000000129208] = 0x00000000899EA638
<129400> 		I-Slot: t2[0x0000000000129200] = 0x0000FFFF395B55BB
...
 
> !bytecodes 0x000000000011C530 
...
    0 aload0getfield 
    1 getfield 10 jdk/internal/ref/PhantomCleanable.list Ljdk/internal/ref/PhantomCleanable;
    4 dup 
    5 astore1 
    6 monitorenter 
    7 aload0getfield 
    8 getfield 10 jdk/internal/ref/PhantomCleanable.list Ljdk/internal/ref/PhantomCleanable;
   11 aload0getfield 
   12 getfield 10 jdk/internal/ref/PhantomCleanable.list Ljdk/internal/ref/PhantomCleanable;
   15 getfield 8 jdk/internal/ref/PhantomCleanable.next Ljdk/internal/ref/PhantomCleanable;
   18 ifacmpne 25
   21 iconst1 
   22 goto 26
   25 iconst0 <-----
   26 aload1 
   27 monitorexit 
   28 returnZ 
   29 astore2 
   30 aload1 
   31 monitorexit 
   32 aload2 
   33 athrow 
...
```

**Comment by babsingh** [comment]
Closing ... the solution https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/750 has been merged and verified through a grinder in https://github.com/eclipse-openj9/openj9/issues/18749#issuecomment-1892386634.
